### ISSUE TITLE
Large files makes neovim high cpu usage and freezes

### ISSUE THREAD

**Issue by menash134** [neovim/neovim#20267]
### Neovim version (nvim -v)

v0.8.0-dev-1117-g7bd4c8e8e

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Linux HPZ2-Siemens 5.15.0-46-generic #49~20.04.1-Ubuntu SMP Thu Aug 4 19:15:44 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Terminal name/version

tmux

### $TERM environment variable

tmux-256color

### Installation

appimage

### How to reproduce the issue

NOTE: I uninstalled vim in my machine. Hence has not cross verified the problem

nvim opening a c file with 13801 lines.
go to insert mode and edit the file
and then see the htop. now cpu usage of nvim is 100%
and it freezes to edit for sometime and recovers.

I profiled the problem in insert mode and attached. my neovim config is  in https://github.com/menash134/nvim.git

[nvim_prof.log](https://github.com/neovim/neovim/files/9611670/nvim_prof.log)

### Expected behavior

zero performance problems in editing large files

### Actual behavior

multiple freezes before adding a single letter in the file.

**Comment by Shougo** [comment]
Please upload the minimal init.vim.
treesitter is very slow in huge C file like this.

**Comment by justinmk** [comment]
yes, don't see anything obvious in the profile log which means the hot path is likely in Lua. Also looks like you have a lot of plugins installed (tagbar, airline), this is noise. Need a minimal repro.

**Comment by menash134** [comment]
Thanks for the quick response and caring about the issue. 
I could pin point to one single plugin the cpu usage and it is 
https://github.com/RRethy/vim-illuminate.git

But indeed I tried with minimal nvim and nvim treesitter and cocnvim. 
The cpu usage shoots upto 30% initially and then it reduces to 10% and then stays at 2% or less. 
when I edit the file, it lags a bit and then it recovers immediately.

This seems to be okay for my use case . The real problem with me having is illuminate plugin. I raised a issue there.

https://github.com/RRethy/vim-illuminate/issues/132

Thanks again for quick support
